Parsimony/Parser.cs

```diff
@@ -0,0 +1,212 @@
+
+            // Consume the token
+            input = input.Skip(1).ToList();
+
+            var optionType = option.Spec.OptionType;
+            var typedOption = typeof(OptionSpec<,>).MakeGenericType(new[] { typeof(T), optionType });
+            var set = typedOption.GetMethod("Set", new[] { typeof(T), optionType });
+
+            if (optionType == typeof(bool))
+            {
+                set.Invoke(option.Spec, new object[] { options, true });
+                option.Set = true;
+
+                if (match.Groups["rest"].Success)
+                    input.Insert(1, $"-{match.Groups["rest"].Value}");
+
+                return new ParserState(new ParseResult<T>(options, arguments), input);
+            }
+
+            string nextToken()
+            {
+                if (input.Count == 0)
+                    // TODO: Custom exception
+                    throw new Exception($"Option -{optionName} requires an argument");
+
+                var value = input[0];
+                input.RemoveAt(0);
+                return value;
+            }
+
+            var valueToken = match.Groups["rest"].Success
+                ? match.Groups["rest"].Value
+                : nextToken();
+
+            var parse = typedOption.GetMethod("Parse", new[] { typeof(string) });
+            var value = parse.Invoke(option.Spec, new object[] { valueToken });
+            set.Invoke(option.Spec, new object[] { options, value });
+            option.Set = true;
+            return new ParserState(new ParseResult<T>(options, arguments), input);
+        }
+
+        private ParserState? ConsumeLongNameOption(ParserState state)
+        {
+            return null;
+        }
+
+        private ParserState? ConsumeArgument(ParserState state)
+        {
+            if (state.Input.Count == 0)
+                return null;
+
+            if (_configuration.DoubleDash && state.Input[0] == "--" || _configuration.OptionsFirst)
+            {
+                return ConsumeAllArguments(state);
+            }
+
+            var options = state.Result.Options;
+            var arguments = state.Result.Arguments.ToList();
+            arguments.Add(state.Input[0]);
+            var input = state.Input.Skip(1);
+
+            return new ParserState(new ParseResult<T>(options, arguments), input);
+        }
+
+        private ParserState? ConsumeAllArguments(ParserState state)
+        {
+            if (state.Input.Count == 0)
+                return null;
+
+            if (_configuration.DoubleDash)
+            {
+                var index = state.Input.IndexOf("--");
+                if (index >= 0)
+                    state.Input.RemoveAt(index);
+            }
+
+            var arguments = state.Result.Arguments.ToList();
+            arguments.AddRange(state.Input);
+            return new ParserState(
+                new ParseResult<T>(state.Result.Options, arguments), Array.Empty<string>());
```

Comment:
Why Array.Empty vs new string[]?

Reply:
its a warning
![image](https://user-images.githubusercontent.com/8669000/78085446-5fa01400-7391-11ea-95e8-cca2694d6fdd.png)

Reply:
Huh. The more you know